hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotMoveTableKeysRequest.java

```diff
@@ -0,0 +1,166 @@
+
+      // If next snapshot is not active then ignore move. Since this could be a redundant operations.
+      if (nextSnapshot != null && nextSnapshot.getSnapshotStatus() != SnapshotInfo.SnapshotStatus.SNAPSHOT_ACTIVE) {
+        throw new OMException("Next snapshot : " + nextSnapshot + " in chain is not active.",
+            OMException.ResultCodes.INVALID_SNAPSHOT_ERROR);
+      }
+
+      String expectedBucketKeyPrefix = omMetadataManager.getBucketKeyPrefix(fromSnapshot.getVolumeName(),
+          fromSnapshot.getBucketName());
+      String expectedBucketKeyPrefixFSO = omMetadataManager.getBucketKeyPrefixFSO(fromSnapshot.getVolumeName(),
+          fromSnapshot.getBucketName());
+
+      // Filter only deleted keys with atlest one keyInfo per key.
+      Set<String> keys = new HashSet<>();
+      List<SnapshotMoveKeyInfos> deletedKeys =
+          moveTableKeysRequest.getDeletedKeysList().stream()
+              .filter(snapshotMoveKeyInfos -> !snapshotMoveKeyInfos.getKeyInfosList().isEmpty())
+              .collect(Collectors.toList());
+      //validate deleted key starts with bucket prefix.[/<volName>/<bucketName>/]
+      for (SnapshotMoveKeyInfos deletedKey : deletedKeys) {
+        if (!deletedKey.getKey().startsWith(expectedBucketKeyPrefix)) {
+          throw new OMException("Deleted Key: " + deletedKey + " doesn't start with prefix " + expectedBucketKeyPrefix,
+              OMException.ResultCodes.INVALID_KEY_NAME);
+        }
+        if (keys.contains(deletedKey.getKey())) {
+          throw new OMException("Duplicate Deleted Key: " + deletedKey + " in request",
+              OMException.ResultCodes.INVALID_REQUEST);
+        } else {
+          keys.add(deletedKey.getKey());
+        }
+      }
+      keys.clear();
+      List<HddsProtos.KeyValue> renamedKeysList = moveTableKeysRequest.getRenamedKeysList().stream()
+          .filter(keyValue -> keyValue.hasKey() && keyValue.hasValue()).collect(Collectors.toList());
+      //validate rename key starts with bucket prefix.[/<volName>/<bucketName>/]
+      for (HddsProtos.KeyValue renamedKey : renamedKeysList) {
+        if (!renamedKey.getKey().startsWith(expectedBucketKeyPrefix)) {
+          throw new OMException("Rename Key: " + renamedKey + " doesn't start with prefix " + expectedBucketKeyPrefix,
+              OMException.ResultCodes.INVALID_KEY_NAME);
+        }
+        if (keys.contains(renamedKey.getKey())) {
+          throw new OMException("Duplicate rename Key: " + renamedKey + " in request",
+              OMException.ResultCodes.INVALID_REQUEST);
+        } else {
+          keys.add(renamedKey.getKey());
+        }
+      }
+      keys.clear();
+      // Filter only deleted dirs with only one keyInfo per key.
+      List<SnapshotMoveKeyInfos> deletedDirs = moveTableKeysRequest.getDeletedDirsList().stream()
+          .filter(snapshotMoveKeyInfos -> snapshotMoveKeyInfos.getKeyInfosList().size() == 1)
```

Comment:
Why do we need to do this filter? `deletedDirs` should only have one `KeyInfo`. So we would be traversing for nothing here.

Reply:
Moved the entire validation to pre execute

Reply:
We would only commit the wrong change only if we send wrong change from `SnapshotDeletingService` itself. Otherwise we know deleteDir will only have one keyInfo.